integrations/kube-agent-updater/cmd/teleport-kube-agent-updater/main.go

```diff
@@ -130,16 +145,76 @@ func main() {
 		os.Exit(1)
 	}
 
-	versionServerURL, err := url.Parse(strings.TrimRight(versionServer, "/") + "/" + versionChannel)
-	if err != nil {
-		ctrl.Log.Error(err, "failed to parse version server URL, exiting")
-		os.Exit(1)
+	// Craft the version getter and update triggers based on the configuration (use RFD-109 APIs, RFD-184, or both).
+	var criticalUpdateTriggers []maintenance.Trigger
+	var plannedMaintenanceTriggers []maintenance.Trigger
+	var versionGetters []version.Getter
+
+	// If the proxy server is specified, we enabled RFD-184 updates
+	// See https://github.com/gravitational/teleport/blob/master/rfd/0184-agent-auto-updates.md#updater-apis
+	if proxyAddress != "" {
+		ctrl.Log.Info("fetching versions from the proxy /find endpoint", "proxy_server_url", proxyAddress)
+
+		proxyClt, err := webclient.NewReusableClient(&webclient.Config{
+			Context:     ctx,
+			ProxyAddr:   proxyAddress,
+			UpdateGroup: updateGroup,
+		})
+		if err != nil {
+			ctrl.Log.Error(err, "failed to create proxy client, exiting")
+			os.Exit(1)
+		}
+
+		// We do a preflight check before starting to know if the proxy is correctly configured and reachable.
+		ctrl.Log.Info("preflight check: ping the proxy server", "proxy_server_url", proxyAddress)
+		pong, err := proxyClt.Ping()
+		if err != nil {
+			ctrl.Log.Error(err, "failed to ping proxy, either the proxy address is wrong, or the network blocks connections to the proxy",
+				"proxy_address", proxyAddress,
+			)
+			os.Exit(1)
+		}
+		ctrl.Log.Info("proxy server successfully pinged",
+			"proxy_server_url", proxyAddress,
+			"proxy_cluster_name", pong.ClusterName,
+			"proxy_version", pong.ServerVersion,
+		)
+
+		versionGetters = append(versionGetters, version.NewProxyVersionGetter("proxy update protocol", proxyClt))
+
+		// In RFD 184, the server is driving the update, so both regular maintenances and
+		// critical ones are fetched from the proxy. Using the same trigger ensures we hit the cache if both triggers
+		// are evaluated and don't actually make 2 calls.
+		proxyTrigger := maintenance.NewProxyMaintenanceTrigger("proxy update protocol", proxyClt)
+		criticalUpdateTriggers = append(criticalUpdateTriggers, proxyTrigger)
+		plannedMaintenanceTriggers = append(plannedMaintenanceTriggers, proxyTrigger)
 	}
-	versionGetter := version.NewBasicHTTPVersionGetter(versionServerURL)
+
+	// If the version server is specified, we enable RFD-109 updates
+	// See https://github.com/gravitational/teleport/blob/master/rfd/0109-cloud-agent-upgrades.md#kubernetes-model
+	if versionServer != "" {
```

Comment:
should this be a if else instead? Having both versionServer and proxyAddress can create problems?

Reply:
We explicitly want to support both at the same time and do a failover based on which one is implemented. That's because we don't know in Helm if the proxy server supports the new updater API. This way we are doing some opportunistic thing by default:
- if no custom version server is specified in the chart, we set both proxy && version server
- if the proxy implements the update protocol, it takes precedence
- else we fall back to the good ol' version server
- we don't change anything if we're a custom version server is set in the chart because that would be a breaking change